### ISSUE TITLE
Convert parameters to destructured object refactor

### ISSUE THREAD

**Issue by cancerberoSgx** [dsherret/ts-morph#598]
Add `convertParamsToDestructuredObject()` method to `ParameteredNode` so any function-like node's parameters can be converted to a destructured object form: `function f(a: number, b: string[]) { }` to `function f({ a, b }: { a: number; b: string[]; }) { }}`. 

Example: 

```ts
sourceFile.getClassOrThrow("C").getMethods().forEach(m => m.convertParamsToDestructuredObject())
```

This refactor is available in 3.4.1 under the "Convert parameters to destructured object" name.

**Comment by dsherret** [comment]
I wonder how difficult it would be for us to implement this ourselves. How does this refactor affect statements that call the method/function?

**Comment by cancerberoSgx** [comment]
> I wonder how difficult it would be for us to implement this ourselves. 

Some time ago I implement this based on ts-simple-ast and it was not so hard but I'm afraid I didn't considered all edge cases like super calls, overrides, overloads, subclasses, etc. 

> How does this refactor affect expressions that call the method/function?

It changes all call expression referring to that function

I was thinking about adding more and more built-in TypeScript code fixes / refactors to the library. I would love to hear your thoughts about it.

TL;DR:

 * I think we should support as many as built-in TypeScript code-fixes/refactors in the library since that would be cheap for us and it will add great value to ts-morph. 
 * I think we should also start thinking about implementing some of these ourself, progressively,  since that is the ultimate scenario for a library like this. It will help us to evaluate current APIs/design.

These are some thoughts related to TypeScript: 

 * TypeScript team is focusing more and more on providing new code-fixes. Some of them are contributions from users. 
 * Once TypeScript team agrees on add a new code fix they need to support it and fix any issue / edge case. I see from their repo that they are taking this relatively seriously by the amount of tests there are for each of the code fix, even the most trivial ones. 
 * in general, the most interesting code fixes, are not easy to implement and have lots of edge cases that need to be supported. 
 * a code fix must not break the code (they should be safe)

And these are some thoughts related to ts-morph:

 * I'm not sure if supporting / implementing these kind of code fixes should be the primary objective of this library. I think it should, but I really wonder what do you think.
 * I we take the road to implement these code fixes ourself then we will need to maintain / tests them which is not a small scope. 
 * On the other hand, if we implement these ourself it will give us the opportunity to refine current APIs and possible design new ones since implementing this kind of stuff is the ultimate scenario for a library like this (and specially now with the new project name). In other words (since my english is not perfect), the navigation part of the library is nice, but IMO what it makes this library unique is the "manipulation" features, and what better way of testing these APIs than developing  code refactor tools like these ? Again I would love to hear your thoughts about this. 
 * I don't think supporting TypeScript built-in code fixes and developing them ourself are exclusive tasks. I think we could to both things, first supporting all of their fixes in the API like we did for OrganizeImports, FixMissingImports, etc (just calling the Language service and applyTextChange , probably forgetting nodes, and then , progressively, replacing these implementations with our own. 
 * In the case we decide to include more TypeScript built-in code fixes, I think we should do it in a more generic way, perhaps not even include a new method on Node/SourceFile, and just support them in in LanguageService?

**Comment by dsherret** [comment]
Ok, I've been thinking about this more..

I think methods like `organizeImports` and `fixUnusedIdentifiers` on source file are ok to forget nodes because they're refactors people can reasonably do as a last step. Refactors like this "Convert parameters to destructured object" aren't reasonable to require as last step and many users might be confused about nodes being forgotten in this scenario.

So basically, I think this refactor should not get first class support and it should only exist as you have it:

```ts
const edits = project.getLanguageService().getEditsForRefactor(sourceFile, {}, param,
                "Convert parameters to destructured object", "Convert parameters to destructured object", {});
if (edits)
    edits.applyChanges();
```

If we're going to have this refactor with first class support, then we should implement it ourselves in a way that nodes won't be forgotten.

**Comment by cancerberoSgx** [comment]
Yes it does, so basically I should remove the method from ParameteredNode and add it to LanguageService, right ? When you say "first class" you are referring to Node subclasses and Project basically right ? Thanks
